Incorrect hover and content assist with lombok

snjeza opened the pull request:
Fixes #1284 
Fixes #1185

datho7561 commented:
Completion on:

```java
...
public class Gr|eeting {
...
```
still doesn't seem to work. The AST we get and use for completion contains all the generated source code, but uses the offset from the non-generated code, so the result is complete nonsense. Perhaps we will need to do something to get a copy of the AST without the generated code?

snjeza commented:
@datho7561 I have updated the PR. I have also updated [testlombok.zip](https://github.com/user-attachments/files/19448617/testlombok.zip) to work with Java 24

datho7561 commented:
Okay. I'm taking another look.

datho7561 commented:
This works well for the hover and completion cases you listed, but this causes some regressions to our completion and search implementations. ~~I think that it's related to the 0 length name AST nodes sometimes being selected by the `NodeFinder.perform()`, which we use in `DOMCompletionContext` to determine what to complete.~~ Sorry, should have looked at the changes before assuming this, I'll try to figure out what's going on.

datho7561 commented:
Example of a completion regression: `CompletionTests18.test402081d`

snjeza commented:
> Example of a completion regression: CompletionTests18.test402081d

~~I'm checking...~~

snjeza committed:
Incorrect hover and content assist with lombok

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

snjeza commented:
test this please

snjeza commented:
@datho7561 I have updated the PR - https://ci.eclipse.org/ls/job/jdt-core-incubator/job/PR-1310/11/

mickaelistria commented:
I don't thinkwe should have some `isGeneratedByLombok` everywhere, but instead just `isGenerated` as other annotation processor can achieve the same.

datho7561 reviewed (approved):
Works a lot better than before.

There are still some edge cases that don't work as expected, and syntax highlighting for a file with lombok annotations doesn't work well:

![a java file with lombok annotations open in eclipse. the text colour of the lombok annotations changes from grey, beige, and black seemingly arbitrarily](https://github.com/user-attachments/assets/503dc9e4-bf14-43cc-a7e5-9c08fb344d5e)

As a "best effort support", I think this is good. I don't think there is a perfect way to handle it.

snjeza commented:
> I don't thinkwe should have some isGeneratedByLombok everywhere, but instead just isGenerated as other annotation processor can achieve the same.

@datho7561 @mickaelistria I have changed the method name. It checks only lombok for now.

mickaelistria commented:
Thanks.
I haven't deeply tested it, but if it doesn't cause regression in tests and @datho7561 likes it, then feel free to merge.

datho7561 commented:
Thank you, Snjezana!
